### PR TITLE
Disable zoom buttons on min/max zoom

### PR DESCRIPTION
**Pull request by danji90**
# How to

Hooks were added in the Zoom component, listening to the map view.
The zoom buttons are disabled respectively if min or max levels are reached.

Successfully tested in trafimage-maps and mapset

# Others

- [x] It's not a hack or at least an unauthorized hack :).
- [x] The images added are optimized.
- [x] Everything in ticket description has been fixed.
- [x] The author of the MR has made its own review before assigning the reviewer.
- [x] IE11 tested.
- [x] Labels applied. if it's a release? a hotfix?
- [x] Tests added.

### CODE REVIEW

File: src/components/Zoom/Zoom.js [hunk 1]
```diff
@@ -110,27 +145,33 @@ function Zoom({
...
     // eslint-disable-next-line react/jsx-props-no-spreading
     <div className="rs-zooms-bar" {...other}>
-      <div
-        role="button"
+      <button
+        type="button"
         tabIndex={0}
-        className="rs-zoom-in"
+        className={`rs-zoom-in${
+          disabledControls.zoomIn ? ' rs-zoom-disabled' : ''
```
Comment (oterral) [inline src/components/Zoom/Zoom.js:152]: use the disabled attribute, no need of a css class for that.

File: src/components/Zoom/Zoom.js [hunk 2]
```diff
@@ -110,27 +145,33 @@ function Zoom({
...
+        }`}
         title={titles.zoomIn}
         onClick={zoomIn}
         onKeyPress={zoomIn}
+        disabled={disabledControls.zoomIn}
```
Comment (oterral) [inline src/components/Zoom/Zoom.js:157]: Why not ?
```suggestion
        disabled={view.getZoom() >= view.getConstrainedZoom(view.getMaxZoom())}
```

File: src/components/Zoom/Zoom.js [hunk 3]
```diff
@@ -91,6 +95,37 @@ function Zoom({
...
+  }, []);
+
+  useEffect(() => {
+    const view = map.getView();
+    // We use the constrained max and min zoom, which is calculated including the viewport size
+    setDisabledControls({
+      zoomIn: view.getZoom() >= view.getConstrainedZoom(view.getMaxZoom()),
+      zoomOut: view.getZoom() <= view.getConstrainedZoom(view.getMinZoom()),
+    });
+  }, [currentZoom]);
```
Comment (oterral) [inline src/components/Zoom/Zoom.js:117]: Can you try with useMEmo:

```
const zoomInDisabled = useMemo(()=> {
return view.getZoom() >= view.getConstrainedZoom(view.getMaxZoom());
}, [view.getZoom()])
```
I don't know if it's work, but itcould save some lines of code

File: src/components/Zoom/Zoom.js [hunk 4]
```diff
@@ -91,6 +95,37 @@ function Zoom({
...
+
+  useEffect(() => {
+    /* Re-add the view listener in case the view was reloaded (e.g. when switching
+     * between certain topics in Trafimage-Maps)
+     */
+    map.on('change:view', () => {
```
Comment (oterral) [inline src/components/Zoom/Zoom.js:123]: Use 'moveend' event

**Review by oterral — CHANGES_REQUESTED** [review]
It could be nice if we could save some line sof code, see my comments

File: src/components/Zoom/Zoom.js (hunk 1 above)
Comment (danji90) [reply src/components/Zoom/Zoom.js]: Done

File: src/components/Zoom/Zoom.js (hunk 2 above)
Comment (danji90) [reply src/components/Zoom/Zoom.js]: Using useMemo

File: src/components/Zoom/Zoom.js (hunk 3 above)
Comment (danji90) [reply src/components/Zoom/Zoom.js:111]: Done

File: src/components/Zoom/Zoom.js (hunk 4 above)
Comment (danji90) [reply src/components/Zoom/Zoom.js]: Done

**Comment by danji90** [comment]
@oterral I applied your suggestions and retested in Trafimage-Maps

File: src/components/Zoom/Zoom.js [hunk 5]
```diff
@@ -90,7 +96,26 @@ function Zoom({
...
+  const zoomInDisabled = useMemo(() => {
+    return (
+      map.getView().getZoom() >=
+      map.getView().getConstrainedZoom(map.getView().getMaxZoom())
+    );
+  }, [currentZoom]);
+
+  const zoomOutDisabled = useMemo(() => {
+    return (
+      map.getView().getZoom() <=
```
Comment (oterral) [inline src/components/Zoom/Zoom.js:108]: you can use currentZoom here
```suggestion
      currentzoom <=
```

File: src/components/Zoom/Zoom.js [hunk 6]
```diff
@@ -90,7 +96,26 @@ function Zoom({
     [map],
   );
 
+  const zoomInDisabled = useMemo(() => {
+    return (
+      map.getView().getZoom() >=
```
Comment (oterral) [inline src/components/Zoom/Zoom.js:101]:
```suggestion
      currentZoom >=
```

File: src/components/Zoom/Zoom.js [hunk 7]
```diff
@@ -90,7 +96,26 @@ function Zoom({
...
+  const zoomOutDisabled = useMemo(() => {
+    return (
+      map.getView().getZoom() <=
+      map.getView().getConstrainedZoom(map.getView().getMinZoom())
+    );
+  }, [currentZoom]);
+
   useEffect(() => {
+    /* Trigger recalculation of disabled zooms */
+    map.on('moveend', () => {
```
Comment (oterral) [inline src/components/Zoom/Zoom.js:115]: you have to unlisten this function in the function returned by the useEffect

**Review by oterral — CHANGES_REQUESTED** [review]
Still 2 minors improvments then you can merge

File: src/components/Zoom/Zoom.js (hunk 5 above)
Comment (danji90) [reply src/components/Zoom/Zoom.js:108]: Tests were failing when I used currentZoom. I'll have another look

File: src/components/Zoom/Zoom.js (hunk 6 above)
Comment (danji90) [reply src/components/Zoom/Zoom.js]: Done

File: src/components/Zoom/Zoom.js (hunk 7 above)
Comment (danji90) [reply src/components/Zoom/Zoom.js]: Done

File: src/components/Zoom/Zoom.test.js [hunk 8]
```diff
@@ -91,4 +93,49 @@ describe('Zoom', () => {
...
+    view: new OLView({ minZoom: 2, zoom: 3 }),
+  });
+  const spy = jest.spyOn(map.getView(), 'setZoom');
+  const wrapper = mount(<Zoom map={map} />);
+  wrapper.find('.rs-zoom-out').first().simulate('click');
+  expect(spy).toHaveBeenCalledTimes(1);
+  act(() => {
+    map.dispatchEvent(new MapEvent('moveend', map));
+  });
+  wrapper.update();
```
Comment (oterral) [inline src/components/Zoom/Zoom.test.js:139]: it worked!!!! :D

**Review by oterral — APPROVED** [review]
lgtm

**Pull request merged**